### ISSUE TITLE
Engine changes for VariantEval

### ISSUE THREAD

**Issue by cmnbroad** [broadinstitute/gatk#5439]
The GATK4 port of GATK3 VariantEval uses a MultiVariantWalker traversal, along with individual `FeatureInput` arguments for evals, knowns, comps, etc., which are all manually merged together as the walker's driving variants. The resulting variants are then manually processed in groups, by start position. Since the tool needs to know the origin of each variant (eval, comp, dbsnp, known, etc.), and since this isn't preserved by the engine, it re-queries the `FeatureContext` for each input to get the same set of variants grouped by source. Since the inputs are typed as `FeatureInput`, this results in all inputs being both consumed and cached twice; once by `MultiVariantDataSource` and once by `FeatureManager`.

Once alternative would be to use a LocusWalker, but that would still require index queries (though the features would be cached), and it would still require manual filtering/aggregation on start position. Proposed fix is to switch the base class to use `MultiVariantWalkerGroupedOnStart`; change the engine to preserve the input source of each variant as proposed in https://github.com/broadinstitute/gatk/pull/4571; and change the input arguments for VariantEval from individual named arguments to tagged feature inputs. This would greatly simplify the initialization code, eliminate redundant reading and caching, and allow the tool to do the input source grouping by just looking at each variant's source field.

**Comment by cmnbroad** [comment]
The tool is currently marked `@BetaFeature` ince the proposed change will require changing the input arguments. Once this issue is resolved, the `@BetaFeature` tag should be removed.

**Comment by cmnbroad** [comment]
@bbimber I have branches where I've started the work for this ticket as well as https://github.com/broadinstitute/gatk/issues/5441. It may be hard to do them separately. They're not ready yet, and I hope to get back to it sometime this month, but just wanted you to know in case you're planning to spend any time on these.

**Comment by bbimber** [comment]
@cmnbroad It's become very apparent that I need to improve the perf of VariantEval, in order to make our VariantQC work better on large WGS VCFs. I dont know where you stand on this, but I'd be willing to pick it up.

I have done virtually zero scoping, but I was also pondering an attempt at making it work w/ Spark.

**Comment by cmnbroad** [comment]
@bbimber Let me see if I can dig up the old branches where I started this work. I'm not sure what state they're in, and I'm sure they're very out of date wrt/master, but they're probably at least worth looking at.

I recall thinking that https://github.com/broadinstitute/gatk/issues/5441 would be quite a lot of work. Have you done any profiling to see where the performance issues actually are for your use case? I'd highly recommend doing that first before embarking on any big changes like this.

**Comment by bbimber** [comment]
@cmnbroad OK, thanks. I need to do more investigation, but my initial thoughts/investigation were two-fold:

1) Our main use-case if our VariantQC tool, which makes several instances of VariantEval in kind of a hacky way and calls apply() on them. Therefore refactoring a VariantEvalEngine out of VariantEval has value on this front, even if tricky.

2) From what I could tell, the worst perf is the iteration pattern. Using something like MultiVariantWalkerGroupedOnStart would be a big savings and avoid re-querying the component VCFs.

**Comment by cmnbroad** [comment]
Both sound right, but it would be comforting to know for sure that the proposed changes will actually fix the problem. I ran a profiler on some of the existing test cases, but they're not substantial enough to make the profiling results interesting.

My previous hacking branch is https://github.com/broadinstitute/gatk/tree/cn_variant_eval_walker. The changes were basically exploratory - I was really just hacking to see what kinds of issues we'd run into. The branch is way behind master so it really needs a rebase.

It would be great it you want to pursue this, but I'd strongly suggest trying to break it up into a series of incremental PRs as much as possible. I haven't looked at these classes for quite a while, but one proposal would be to start with a PR that just does #5441, then one that factors out the eval engine so it can be reused, and then another to change the VariantEval base class.

**Comment by bbimber** [comment]
I will do my best to split apart. My initial reaction was actually that switching to MultiVariantWalkerGroupedOnStart was going to be the most discrete and separable fix, but i will see.

**Comment by bbimber** [comment]
@cmnbroad I started down this road. i wanted to make sure i follow your reasoning on some of this. I think you propose to change the tool arguments such that each input VCF is tagged by name (like -V:eval vcf1.vcf.gz -V:comp vcf2.vcf.gz), instead of different argument names. This is paired with a change to set the 'source' on each VariantContext to match the name of the source feature context. Unless I'm missing something, this basically makes everything identified by strings, with no direct FeatureInput <-> VariantContext reference, right?  

Presumably, MultiVariantWalkerGroupedOnStart could implement something like:

    protected Map<FeatureInput<VariantContext>, List<VariantContext>> groupVariantsByFeatureInput(List<VariantContext> variants) {
        Map<String, FeatureInput<VariantContext>> sourceMap = new HashMap<>();
        getDrivingVariantsFeatureInputs().forEach(x -> sourceMap.put(x.getName(), x));
        
        Map<FeatureInput<VariantContext>, List<VariantContext>> ret = new HashMap<>();
        variants.forEach(vc -> {
            FeatureInput<VariantContext> fi = sourceMap.get(vc.getSource());
            if (fi == null) {
                 //possibly throw?                
            }
            
            List<VariantContext> l = ret.getOrDefault(fi, new ArrayList<>());
            l.add(vc);
            ret.put(fi, l);
        });        
        
        return ret;
    }

**Comment by cmnbroad** [comment]
@bbimber Yes, I think this is along the lines of what I was previously thinking.
